File: src/lib/jev/fp/cospan.ts
import type { Color } from "../colors.ts";
import { err, ok, type Result } from "./result.ts";

export type Port = { readonly id: string; readonly color: Color; readonly label: string };
export type System = { readonly id: string; readonly name: string; readonly ports: readonly Port[] };

export type GlueError = { readonly kind: "color-mismatch"; readonly color: Color };

/** A → I ← B. Apex I is the glued color. */
export type Cospan = {
  readonly left: System;
  readonly right: System;
  readonly apex: Color;
  readonly leftIndex: number;
  readonly rightIndex: number;
};

/** Pushout of the cospan: leftover outer ports, identified glue remembered. */
export type Pushout = {
  readonly system: System;
  readonly identified: Color;
};

export function encodePorts(ports: readonly Port[]): number[] {
  const order: Color[] = ["entity", "concept", "idea", "evidence", "action"];
  return ports.map((p) => order.indexOf(p.color));
}

export function cospanOf(left: System, right: System, color: Color): Result<Cospan, GlueError> {
  const leftIndex = left.ports.findIndex((p) => p.color === color);
  const rightIndex = right.ports.findIndex((p) => p.color === color);
  if (leftIndex < 0 || rightIndex < 0) return err({ kind: "color-mismatch", color });
  return ok({ left, right, apex: color, leftIndex, rightIndex });
}

/** Discrete pushout — twin of Rust `pushout_ports`. */
export function pushout(c: Cospan): Pushout {
  const rest = [...c.left.ports.filter((_, i) => i !== c.leftIndex), ...c.right.ports.filter((_, i) => i !== c.rightIndex)];
  return {
    identified: c.apex,
    system: {
      id: `${c.left.id}+${c.right.id}`,
      name: `${c.left.name} ⋈ ${c.right.name}`,
      ports: rest,
    },
  };
}

export function glue(left: System, right: System, color: Color): Result<Pushout, GlueError> {
  const c = cospanOf(left, right, color);
  return c.ok ? ok(pushout(c.value)) : c;
}

File: crates/jev-kernel/src/cospan.rs
//! Cospans of colored interfaces.
//!
//! A cospan is A → I ← B. The apex I is the glue (shared ports).
//! Composition of port-plugging systems is the pushout of that cospan:
//! identify the matching color and emit the leftover outer ports.

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum GlueError {
    ColorMismatch,
}

/// Discrete pushout: drop the first matching `color` on each foot, concatenate the rest.
pub fn pushout_ports(left: &[u8], right: &[u8], color: u8) -> Result<Vec<u8>, GlueError> {
    let li = left.iter().position(|&p| p == color).ok_or(GlueError::ColorMismatch)?;
    let ri = right.iter().position(|&p| p == color).ok_or(GlueError::ColorMismatch)?;
    let mut out = Vec::with_capacity(left.len() + right.len() - 2);
    out.extend(left.iter().enumerate().filter(|(i, _)| *i != li).map(|(_, p)| *p));
    out.extend(right.iter().enumerate().filter(|(i, _)| *i != ri).map(|(_, p)| *p));
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn glue_entity() {
        // left: entity, evidence   right: entity, concept
        let out = pushout_ports(&[0, 3], &[0, 1], 0).unwrap();
        assert_eq!(out, vec![3, 1]);
    }

    #[test]
    fn mismatch_is_err() {
        assert_eq!(pushout_ports(&[0, 3], &[1, 4], 4), Err(GlueError::ColorMismatch));
    }
}
